stan/math/prim/mat/functor/mpi_parallel_call.hpp

```diff
@@ -0,0 +1,488 @@
+
+    const size_type num_jobs = job_dims[0];
+    const size_type num_job_params = num_jobs == 0 ? 0 : job_dims[1];
+
+    const vector_d shared_params_dbl = value_of(shared_params);
+    matrix_d job_params_dbl(num_job_params, num_jobs);
+
+    for (int j = 0; j < num_jobs; ++j)
+      job_params_dbl.col(j) = value_of(job_params[j]);
+
+    setup_call(shared_params_dbl, job_params_dbl, x_r, x_i);
+  }
+
+  // called on remote sites
+  mpi_parallel_call() : combine_() {
+    if (rank_ == 0)
+      throw std::runtime_error("problem sizes must be defined on the root.");
+
+    setup_call(vector_d(), matrix_d(), std::vector<std::vector<double>>(),
+               std::vector<std::vector<int>>());
+  }
+
+  static void distributed_apply() {
+    // call constructor for the remotes
+    mpi_parallel_call<call_id, ReduceF, CombineF> job_chunk;
+
+    job_chunk.reduce_combine();
+  }
+
+  // all data is cached and local parameters are also available
+  result_t reduce_combine() {
+    const std::vector<int>& job_chunks = cache_chunks::data();
+    const int num_jobs = sum(job_chunks);
+
+    // id of first job out of all
+    int start_job = 0;
+    for (std::size_t n = 0; n < rank_; ++n)
+      start_job += job_chunks[n];
+
+    const int num_local_jobs = local_job_params_dbl_.cols();
+    int local_outputs_per_job = num_local_jobs == 0 ? 0 : num_outputs_per_job_;
+    matrix_d local_output(
+        local_outputs_per_job == -1 ? 0 : local_outputs_per_job,
+        num_local_jobs);
+    std::vector<int> local_f_out(num_local_jobs, -1);
+
+    typename cache_x_r::cache_t& local_x_r = cache_x_r::data();
+    typename cache_x_i::cache_t& local_x_i = cache_x_i::data();
+
+    // check if we know already output sizes
+    if (cache_f_out::is_valid()) {
+      typename cache_f_out::cache_t& f_out = cache_f_out::data();
+      int num_outputs = 0;
+      for (int j = start_job; j < start_job + num_local_jobs; ++j)
+        num_outputs += f_out[j];
+      local_output.resize(Eigen::NoChange, num_outputs);
+    }
+
+    int local_ok = 1;
+    try {
+      for (int i = 0, offset = 0; i < num_local_jobs;
+           offset += local_f_out[i], ++i) {
+        const matrix_d job_output
+            = ReduceF()(local_shared_params_dbl_, local_job_params_dbl_.col(i),
+                        local_x_r[i], local_x_i[i], 0);
+        local_f_out[i] = job_output.cols();
+
+        if (local_outputs_per_job == -1) {
+          local_outputs_per_job = job_output.rows();
+          local_output.conservativeResize(local_outputs_per_job,
+                                          Eigen::NoChange);
+        }
+
+        if (local_output.cols() < offset + local_f_out[i])
+          local_output.conservativeResize(Eigen::NoChange,
+                                          2 * (offset + local_f_out[i]));
+
+        local_output.block(0, offset, local_output.rows(), local_f_out[i])
+            = job_output;
+      }
+    } catch (const std::exception& e) {
+      local_ok = 0;
```

Comment:
I think I doced this in the doxygen comment above. See Note 1 there. Is that sufficient? I could add a comment which refers to note 1. The thing is that the exception must be thrown on the root only. Otherwise things get out of sync which leads to a catastrophe (undefined behavior).